tcms/issuetracker/bitbucket.py

```diff
@@ -0,0 +1,140 @@
+
+    def add_comment(self, issue_id, comment):
+        url = '{0}{1}{2}{3}'.format(self.endpoint_url, '/issues/', issue_id, '/comments/')
+        return self._request("POST", url, headers=self.headers, auth=self.auth, json=comment)
+
+    def get_comments(self, issue_id):
+        url = '{0}{1}{2}{3}'.format(self.endpoint_url, '/issues/',
+                                    issue_id, '/comments?sort=-updated_on')
+        return self._request("GET", url, headers=self.headers, auth=self.auth)
+
+    @staticmethod
+    def _request(method, url, **kwargs):
+        return requests.request(method, url, **kwargs).json()
+
+    @staticmethod
+    def _construct_endpoint_url(api_version, url):
+        splitted_url = url.replace('https://', '').split('/')
+        base_url = "https://api.bitbucket.org"
+        workspace = splitted_url[1]
+        repository = splitted_url[2]
+        endpoint_url = "{0}/{1}/{2}/{3}/{4}".format(base_url,
+                                                    api_version,
+                                                    'repositories',
+                                                    workspace,
+                                                    repository)
+        return endpoint_url
+
+
+class BitBucketThread(IntegrationThread):
+    """
+        Execute BitBucket code in a thread!
+
+        Executed from the IssueTracker interface methods.
+    """
+
+    def post_comment(self):
+        comment_body = {"content": {"raw": self.text()}}
+        self.rpc.add_comment(self.bug_id, comment_body)
+
+
+class BitBucket(IssueTrackerType):
+    """
+        Support for BitBucket. Requires:
+
+        :base_url: - URL to a BitBucket project for which we're going to report issues
+        e.g. https://bitbucket.org/{workspace}/{repository}
+        :api_username: BitBucket Username
+        :api_password: BitBucket App Password
+
+        .. note::
+
+            You can leave the ``api_url`` field blank because
+            the integration code doesn't use them!
+    """
+    it_class = BitBucketThread
+
+    def _rpc_connection(self):
+        # NOTE: BitBucket 'app password' is required as api_password
+        # NOTE: BitBucket 'username' (not e-mail address) is required as api_username
+        return BitBucketAPI(self.bug_system.base_url,
+                            api_username=self.bug_system.api_username,
+                            api_password=self.bug_system.api_password)
+
+    def is_adding_testcase_to_issue_disabled(self):
+        return not (self.bug_system.base_url and self.bug_system.api_password)
+
+    def report_issue_from_testexecution(self, execution, user):
+        """
+            BitBucket creates the Issue with Title and Description
+        """
+
+        data = {
+            'title': 'Failed test: %s' % execution.case.summary,
+            'kind': 'bug',
+            'priority': 'major',
+            'content': {'raw': self._report_comment(execution)},
```

Comment:
Can we send raw markdown directly to BitBucket ?

Reply:
Currently it looks like this in issue description, when its directly send;

![image](https://user-images.githubusercontent.com/8937774/98235090-65d32000-1f72-11eb-942d-7f51e967956f.png)

And in comments;

![image](https://user-images.githubusercontent.com/8937774/98235141-797e8680-1f72-11eb-965b-e31425c352e0.png)

Reply:
It looks like newlines aren't rendered properly but otherwise lists and bold text seems fine, so no need to call markdown2html() here.

Reply:
Can you figure out what's up with those newlines ?

Reply:
Seems like BitBucket requires double `\n\n` and does not recognize `\r`.
New lines are added when i add `.replace('\r\n', '\n\n')` for description and `.replace('\n','\n\n')` in comments.

Description becomes;
![image](https://user-images.githubusercontent.com/8937774/98333940-da5d9b80-2012-11eb-8782-58bccf5dc5ae.png)

Comment becomes;
![image](https://user-images.githubusercontent.com/8937774/98333961-e5b0c700-2012-11eb-8a35-6ad8a6ea7343.png)

Reply:
The \r\n behavior is odd but it does tell us something. What's your development system - is that a Windows machine or a Linux one ?

Reply:
Linux Mint

Reply:
On Linux there shouldn't be any `\r` characters as far as I know. I will poke around this problem myself as well.

Reply:
So found out what's going on here. For some reason BitBucket has an inconsistency in what characters they need for newlines.  From https://bitbucket.org/kiwitcms/integration/issues/14/testing-newlines:

So their editor adds double `\r\n` for issues description

```
>>> pprint(api.get_issue(14))
{'assignee': None,
 'component': None,
 'content': {'html': '<p>I am testing newlines support here.</p>\n'
                     '<p>This is on a another line, but seemingly there is 1 '
                     'more empty line above. </p>\n'
                     '<p>This is <strong>bold</strong> text.</p>\n'
                     '<hr />\n'
                     '<p>This is under the divider b/c I have no idea if this '
                     'editor is markdown or not.</p>',
             'markup': 'markdown',
             'raw': 'I am testing newlines support here.\r\n'
                    '\r\n'
                    'This is on a another line, but seemingly there is 1 more '
                    'empty line above. \r\n'
                    '\r\n'
                    'This is **bold** text.\r\n'
                    '\r\n'
                    '---\r\n'
                    '\r\n'
                    'This is under the divider b/c I have no idea if this '
                    'editor is markdown or not.',
             'type': 'rendered'},
```

But for comments it uses a double `\n` instead:

```
>>> pprint(api.get_comments(14))
{'page': 1,
 'pagelen': 20,
 'size': 1,
 'values': [{'content': {'html': '<p>Adding a comment to check-out '
                                 'newlines.</p>\n'
                                 '<p>This is on a separate line</p>\n'
                                 '<hr />\n'
                                 '<p>This is under the divider.</p>',
                         'markup': 'markdown',
                         'raw': 'Adding a comment to check-out newlines.\n'
                                '\n'
                                'This is on a separate line\n'
                                '\n'
                                '---\n'
                                '\n'
                                'This is under the divider.',
                         'type': 'rendered'},
```

These have been created manuall via the browser.

Then see my last commit at:
https://github.com/kiwitcms/Kiwi/pull/2077/commits

- when opening a new issue the content already has 2 newline characters which we convert to `\r\n` each. 
- when adding comments though the text content only has a single `\n` character which needs to be doubled. 
- There are no `\r` characters in the input content at all.

Reply:
Thank you for investigating!

I have applied the changes.